File: ois/ois20/round4/ois11_round4_solutions.tar(1)/islands.pas
{ NOTE: it is recommended to use this even if you don't understand the following code. }

{ constraints }
const
    MAXN = 1000;
    dr : array[0..3] of longint = (-1, 1, 0, 0);
    dc : array[0..3] of longint = (0, 0, 1, -1);

{ input data }
var
    R, C, i, j  : longint;
    count: longint;
    M           : array[0..MAXN-1, 0..MAXN-1] of longint;

procedure dfs (rr, cc : longint);
var
    k : longint;
begin
    if (rr < 0) or (rr >= R) or (cc < 0) or (cc >= C) or (M[rr, cc] = 0) then
        exit;
    M[rr, cc] := 0;
    for k:=0 to 3 do
    begin
        dfs(rr+dr[k], cc+dc[k]);
    end;
end;

begin
{
    uncomment the following lines if you want to read/write from files
    assign(input,  'input.txt');  reset(input);
    assign(output, 'output.txt'); rewrite(output);
}

    readln(R, C);
    for i:=0 to R-1 do
    begin
        for j:=0 to C-1 do
            read(M[i,j]);
        readln();
    end;

    for i:=0 to R-1 do
    begin
        dfs(i, 0);
        dfs(i, C-1);
    end;

    for j:=0 to C-1 do
    begin
        dfs(0, j);
        dfs(R-1, j);
    end;

    count := 0;
    for i:=0 to R-1 do
    begin
        for j:=0 to C-1 do
        begin
            if (M[i,j] = 1) then
            begin
                dfs(i, j);
                count := count + 1;
            end;
        end;
    end;

    { insert your code here }

    writeln(count); { print result }
end.

File: ois/ois20/round4/ois11_round4_solutions.tar(1)/islands.py
#!/usr/bin/env pypy3
# NOTE: it is recommended to use this even if you don't understand the following code.

# input data
R, C = map(int, input().strip().split())
M = [None] * R
for i in range(R):
    M[i] = list(map(int, input().strip().split()))

def valid(r, c):
    global R
    global C
    return (0 <= r < R) and (0 <= c < C)

def dfs(r, c):
    global M
    if not valid(r, c) or M[r][c] == 0:
        return
    queue = [(r, c)]
    M[r][c] = 0
    while len(queue) > 0:
        r, c = queue.pop()
        for i, j in zip([-1, 1, 0, 0], [0, 0, 1, -1]):
                if not valid(r+i, c+j) or M[r+i][c+j] == 0:
                    continue
                M[r+i][c+j] = 0
                queue.append((r+i, c+j))

for i in range(R):
    dfs(i, 0)
    dfs(i, C-1)

for j in range(C):
    dfs(0, j)
    dfs(R-1, j)

count = 0
for i in range(R):
    for j in range(C):
        if M[i][j] == 1:
            dfs(i, j)
            count += 1

print(count)  # print the result
